### PR TITLE
[HttpClient] ZIO & fs2 - Switch to non-blocking decompression

### PR DESCRIPTION
**Pull request by ghostbuster91**
Before submitting pull request:
- [x] Check if the project compiles by running `sbt compile`
- [x] Verify docs compilation by running `sbt compileDocs`
- [x] Check if tests pass by running `sbt test`
- [x] Format code by running `sbt scalafmt`

### CODE REVIEW

File: httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientFutureBackend.scala [hunk 1]
```diff
@@ -37,28 +47,44 @@ class HttpClientFutureBackend private (
...
+
+  override protected def publisherToBody(p: Publisher[util.List[ByteBuffer]]): InputStream = {
+    val subscriber = new InputStreamSubscriber
+    p.subscribe(subscriber)
+    subscriber.inputStream
+  }
 }
 
 object HttpClientFutureBackend {
+  type FutureEncodingHandler = EncodingHandler[InputStream]
```
Comment (adamw) [inline httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientFutureBackend.scala:81]: `InputStreamEncodingHandler`? This has nothing to do w/ Futures?

File: httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientFutureBackend.scala [hunk 2]
```diff
@@ -91,3 +117,100 @@ object HttpClientFutureBackend {
   def stub(implicit ec: ExecutionContext = ExecutionContext.global): SttpBackendStub[Future, WebSockets] =
     SttpBackendStub(new FutureMonad())
 }
+
+// based on org.asynchttpclient.request.body.generator.ReactiveStreamsBodyGenerator.SimpleSubscriber
+private[httpclient] class InputStreamSubscriber extends Subscriber[java.util.List[ByteBuffer]] {
```
Comment (adamw) [inline httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientFutureBackend.scala:122]: maybe move to a separate file?

File: httpclient-backend/src/main/scala/sttp/client/httpclient/package.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package sttp.client
+
+import java.nio.ByteBuffer
+
+package object httpclient {
```
Comment (adamw) [inline httpclient-backend/src/main/scala/sttp/client/httpclient/package.scala:5]: maybe move to sttp.client.internal? There's some bytebuffer-related util funcitons already

File: httpclient-backend/src/main/scala/sttp/client/httpclient/InputStreamBodyFromResponseAs.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import java.io.{BufferedInputStream, ByteArrayInputStream, FileInputStream, InputStream}
+
+import sttp.client.internal.{BodyFromResponseAs, FileHelpers, SttpFile}
+import sttp.client.ws.{GotAWebSocketException, NotAWebSocketException}
+import sttp.client.{ResponseMetadata, WebSocketResponseAs}
+import sttp.monad.MonadError
+import sttp.monad.syntax.MonadErrorValueOps
+import sttp.ws.WebSocket
+
+abstract class InputStreamBodyFromResponseAs[F[_], S](implicit
```
Comment (adamw) [inline httpclient-backend/src/main/scala/sttp/client/httpclient/InputStreamBodyFromResponseAs.scala:12]: this is unused?

File: httpclient-backend/monix/src/main/scala/sttp/client/httpclient/monix/HttpClientMonixBackend.scala [hunk 5]
```diff
@@ -45,17 +56,12 @@ class HttpClientMonixBackend private (
         monad.eval(BodyPublishers.fromPublisher(FlowAdapters.toFlowPublisher(stream.toReactivePublisher)))
     }
 
-  override protected val bodyFromHttpClient: BodyFromHttpClient[Task, MonixStreams] =
-    new BodyFromHttpClient[Task, MonixStreams] {
+  override protected val bodyFromHttpClient: BodyFromHttpClient[Task, MonixStreams, InputStream] =
+    new InputStreamBodyFromHttpClient[Task, MonixStreams] {
+      override def inputStreamToStream(is: InputStream): Task[(streams.BinaryStream, () => Task[Unit])] =
+        monad.error(new IllegalStateException("Streaming is not supported"))
```
Comment (adamw) [inline httpclient-backend/monix/src/main/scala/sttp/client/httpclient/monix/HttpClientMonixBackend.scala:62]: it's not?

Comment (ghostbuster91) [reply httpclient-backend/monix/src/main/scala/sttp/client/httpclient/monix/HttpClientMonixBackend.scala]: Ah copy-pasting, fixed :)

**Comment by adamw** [comment]
Thanks! Released in [3.0.0-RC3](https://github.com/softwaremill/sttp/releases/tag/v3.0.0-RC3)

**Pull request merged**